[Orders with Coupons M7] Product discount components

iamgabrielma opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #10837

## Description
This PR adds the updated "Discount View"  to a product, which involves:
- Updating the design of the existing `CollapsibleProductRowCard` to show an "Add discount" button as well as the discount after has been added.
- Adding a new `ProductDiscountView` and `DiscountLineDetailsView`, where the discount calculation and application is performed. This substitutes the previous modal view. 

As with the collapsable product row, the design is not 1:1 with the [proposed design](https://www.figma.com/file/FTOWxedKUyDbBdOrxRfMyI/Coupons-in-Order-Creation?type=design&node-id=14-10245&mode=design&t=Tn0eG6fym2kCjRo7-0), but gets as close as possible. Items I left outside

<img width="275" alt="Screenshot 2023-10-09 at 16 04 09" src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/6f44bbc6-66b2-4c8a-88b1-0e4d82256a6e">

- There is no percentage discount calculation in this view. We just show the discount and the total discounted amount.
- The unique button to switch between fixed and percentage is two buttons for simplicity.

## Caveats:
 - We kept the changes to `FeeOrDiscountLineDetailsViewModel` to the minimum, since this part of the project conflicts with Custom Amounts ( pdfdoF-3Sp-p2 cc: @toupper ). Ideally we'll want to circle back and decouple entirely fees from discounts.
 - Along with the previous point, there are known bugs with the existing discounts implementation (for example being able to enter positive discounts, or being able to discount more than the total price of the product. We're not addressing these with this PR, but are logged in https://github.com/woocommerce/woocommerce-ios/issues/10828 and https://github.com/woocommerce/woocommerce-ios/issues/5143)

![Simulator Screen Recording - iPhone 15 - 2023-10-09 at 15 48 41](https://github.com/woocommerce/woocommerce-ios/assets/3812076/1138d02a-9454-410b-ac67-0231287c8579)

## Testing instructions
* Switch `ordersWithCouponsM6` feature flag to true
* Go to `Orders` > `+` > `+ Add Products` > Add a product to your order, expand the chevron.
* Observe that:
  * ` +Add discount` button is available
  * Tapping on it opens the new product discount view
  * We can enter either a fixed or a percentage amount of discount. Once we do so, the calculation happens on-screen. Note that here is the bug where we can add a discount bigger than the total price of the product. This also renders a `Remove Discount` button.
  * Tapping `Add` adds the discount to the product and dismisses the view. 
  * Now we can see a `Discount` + pencil icon button as well as the discount added to the product. Tapping again allows to edit or remove the discount.

| 1 | 2 | 3 | 4 |
|--------|--------|--------|--------|
| <img width="295" alt="Screenshot 2023-10-09 at 16 12 46" src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/2f47dca7-04e5-41a0-a53d-de37a4898350"> | <img width="292" alt="Screenshot 2023-10-09 at 16 13 10" src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/f3c7b9a8-cea7-48ea-a339-dc98cd451689"> | <img width="293" alt="Screenshot 2023-10-09 at 16 13 29" src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/5f153533-bebf-4e34-8099-c164164f4a4d"> | <img width="296" alt="Screenshot 2023-10-09 at 16 13 49" src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/1e594582-f36b-44f9-b279-3a71f9c482d1"> |

iamgabrielma committed:
Add onAddDiscount callback to collapsible component

iamgabrielma committed:
Discount button style and tooltip. Localization.

iamgabrielma committed:
Rename LegacyProductInOrder

iamgabrielma committed:
New ProductInOrder & DiscountLineDetailsView views

iamgabrielma committed:
Merge branch 'trunk' into issue/10837-order-ux-m7-add-discounts

# Conflicts:
#	WooCommerce/Classes/ViewRelated/Orders/Order Creation/CollapsibleProductRowCard.swift
#	WooCommerce/Classes/ViewRelated/Orders/Order Creation/OrderForm.swift

iamgabrielma committed:
Re-add debug param removed on merge conflict

iamgabrielma committed:
Rename ProductDiscountView

iamgabrielma committed:
Switch fixed-percentage discount viewmodel-based

iamgabrielma committed:
Add components to ProductDiscountView

iamgabrielma committed:
Add swiftui components

iamgabrielma committed:
Expose productrow price. Calculate price post-discount

iamgabrielma committed:
Discount data between productrow and discountview

iamgabrielma committed:
Revert renaming

iamgabrielma committed:
Move layout props to private extension

iamgabrielma committed:
Extract DiscountLineDetailsView to own file

iamgabrielma committed:
Add localized strings

iamgabrielma committed:
Simplify input field placeholders

iamgabrielma committed:
Add icon to editable discount. Localized strings

iamgabrielma committed:
Update Picker with Button components

iamgabrielma committed:
Render discount line when discount not empty

iamgabrielma committed:
Hide hasDiscount implementation details

iamgabrielma committed:
Hide discount input entered implementation details

iamgabrielma committed:
Remove coalescing empty prices. Add minus sign.

iamgabrielma committed:
Refactor calculatePriceAfterDiscount

joshheald commented:
A bug I've noticed that I can't see logged anywhere:

1. Tap Add discount, set an amount, and tap `Add`, the modal is dismissed and the discount shows on the Order Details.
2. Tap the pencil quickly, to change the amount
3. The add discount screen is shown again, not the edit discount view.

https://github.com/woocommerce/woocommerce-ios/assets/2472348/154387ef-3f67-4808-af74-b70ddf8ca6e5

joshheald commented:
A couple of layout issues I spotted:

1. There's no gap between the edge of the screen and the segmented control
2. The product name is centred when it should be leading aligned
3. The `×` symbol looks huge. It's from another PR but I think it'd be better if you pasted the multiplication symbol into the code and used a localized format string rather than using an image. 

None need to be fixed in this PR if it's not convenient but I wanted to mention them.
![annotated layout](https://github.com/woocommerce/woocommerce-ios/assets/2472348/a9a5a0a9-ca2d-4cd2-abf9-d1367c5b4f6b)

joshheald commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/CollapsibleProductRowCard.swift:
```diff
@@ -47,6 +52,49 @@ struct CollapsibleProductRowCard: View {
+                        }, label: {
+                            HStack {
+                                Text(Localization.discountLabel)
+                                Image(uiImage: .pencilImage)
+                                    .resizable()
+                                    .frame(width: Layout.iconSize, height: Layout.iconSize)
```
Idle thought – perhaps we should have a simple view for our GridIcons that neatens this up? Doesn't really need to be in this PR but this boilerplate is a bit of a paper cut compared to using SFSymbols (which I still think are better!)

iamgabrielma replied:
That sounds a good idea, we seem to redo this view a lot and it's indeed some boilerplate that could be reduced. I'll log this on https://github.com/woocommerce/woocommerce-ios/issues/10863 to keep track of it. I can take care of it as part of the project or separately.

joshheald commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/CollapsibleProductRowCard.swift:
```diff
@@ -47,6 +52,49 @@ struct CollapsibleProductRowCard: View {
+                        })
+                        Spacer()
+                        if let discountLabel = viewModel.discountLabel {
+                            Text(minusSign + discountLabel)
```
Will this localize properly for rtl languages? I'm not sure but I don't think it will. Perhaps using a number formatter to make it negative would be better?

iamgabrielma replied:
That's an interesting one, I thought the symbol should be on the opposite side, but that doesn't seem to be the case:

```
Even though the language is read from right to left, the numerical and financial contexts such as prices, especially involving arithmetic symbols, maintain a left-to-right order. This is recognized and understood in RTL languages and cultures.
```

We seem to not deal with this correctly across the app, but I would need to ask/test it further to confirm all cases. This is a quick test of how refunds show with an RTL currency, just to show a different view from this PR. We seem to show currency symbols inconsistently when there's RTL, and the `-` appears in the incorrect place as well:
<img width=500 src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/5142c9c6-7924-4db3-9d7d-36c3721789ff">

Since the value from `discountLabel` is already currency-formatted (but as an absolute value) I tested a couple of approaches like using the existing `isNegative` property, but this still renders the symbol in the wrong place, since changes the position of the `-` symbol either in the view or in the CurrencyFormatter. I've opened a different issue here to investigate further: https://github.com/woocommerce/woocommerce-ios/issues/10913

joshheald commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/CollapsibleProductRowCard.swift:
```diff
@@ -47,6 +52,49 @@ struct CollapsibleProductRowCard: View {
+                        Spacer()
+                        if let discountLabel = viewModel.discountLabel {
+                            Text(minusSign + discountLabel)
+                                .foregroundColor(.green)
```
There's probably a ColorStudio color we should be using here instead of the default green

iamgabrielma replied:
There is! Fixed: 69f9ef08340428f135b31128be1f3a8dc6823843

joshheald commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/ProductsSection/DiscountLineDetailsView.swift:
```diff
@@ -0,0 +1,136 @@
+    }
+
+    var body: some View {
+        VStack(alignment: .leading, spacing: .zero) {
+            Text(Localization.fixedPriceDiscountLabel)
+                .renderedIf(viewModel.feeOrDiscountType == .fixed)
+                .padding()
+            Text(Localization.percentagePriceDiscountLabel)
+                .renderedIf(viewModel.feeOrDiscountType == .percentage)
+                .padding()
+            HStack {
+                inputFixedField
+                    .renderedIf(viewModel.feeOrDiscountType == .fixed)
+                inputPercentageField
+                    .renderedIf(viewModel.feeOrDiscountType == .percentage)
+                HStack(spacing: 0) {
+                    Button(viewModel.currencySymbol) {
+                        viewModel.feeOrDiscountType = .fixed
+                    }
+                    .buttonStyle(isPrimary: viewModel.feeOrDiscountType == .fixed)
+                    .fixedSize(horizontal: true, vertical: false)
+                    Button(viewModel.percentSymbol) {
+                        viewModel.feeOrDiscountType = .percentage
+                    }
+                    .buttonStyle(isPrimary: viewModel.feeOrDiscountType == .percentage)
+                    .fixedSize(horizontal: true, vertical: false)
+                }
+            }
+        }
```
There's a lot of checks for `.fixed` or `.percentage` here. Have you tried making two separate views, one for each, and switching between them with a top level `if`? It would probably be easier to read, and I _think_ the transitions should still work correctly... but they might not.

iamgabrielma replied:
I agree there's room for improvement here. I might also be able to simplify further once we get rid of the "fees" part of the model. Logged on https://github.com/woocommerce/woocommerce-ios/issues/10915

joshheald commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/ProductsSection/DiscountLineDetailsView.swift:
```diff
@@ -0,0 +1,136 @@
+        .padding([.leading, .trailing], Layout.padding)
+        .overlay {
+            RoundedRectangle(cornerRadius: Layout.frameCornerRadius)
+                .inset(by: 0.25)
```
perhaps this should be in the Layout enum?

iamgabrielma replied:
Yes! Fixed: ea5b5b3caad410eda9e06624f5b77458d44dfdd5

joshheald commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/ProductsSection/DiscountLineDetailsView.swift:
```diff
@@ -0,0 +1,136 @@
+        .padding([.leading, .trailing], Layout.padding)
+        .overlay {
+            RoundedRectangle(cornerRadius: Layout.frameCornerRadius)
+                .inset(by: 0.25)
```
as above

iamgabrielma replied:
ea5b5b3caad410eda9e06624f5b77458d44dfdd5

joshheald commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/ProductsSection/ProductDiscountView.swift:
```diff
@@ -0,0 +1,134 @@
+                .padding()
+                .overlay {
+                    RoundedRectangle(cornerRadius: Layout.frameCornerRadius)
+                        .inset(by: 0.25)
```
Another one for the Layout enum

iamgabrielma replied:
ea5b5b3caad410eda9e06624f5b77458d44dfdd5

joshheald commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/ProductsSection/ProductDiscountView.swift:
```diff
@@ -0,0 +1,134 @@
+                    DiscountLineDetailsView(viewModel: discountViewModel)
+                    HStack {
+                        Image(systemName: "arrow.turn.down.right")
+                            .foregroundColor(.gray)
+                        Text(Localization.discountLabel)
+                            .foregroundColor(.gray)
```
Text normally has a semantic colour – I think it would be `.secondary` here? Both work, even in dark mode, so it's not a big problem, but it's not idiomatic.

iamgabrielma replied:
Good point, this went under the radar. Fixed (along similar ones) here: 8810531bc04ad2ef6f05396e4dafa2989255f12b

joshheald reviewed (approved):
Works as described, with the exception of the quick-edit bug described in my previous comment.

A few minor suggestions around the code.

I've approved it with the various known issues since it's feature flagged.

iamgabrielma committed:
Add trailing padding to discount selector

iamgabrielma committed:
Add leading aligment to product in discount view

iamgabrielma committed:
Update discount text with correct ColorStudio

iamgabrielma committed:
Declare constant in private Layout enum

iamgabrielma committed:
Update secondary text with semantic colour

iamgabrielma committed:
Update navigation title on add-update discount

iamgabrielma committed:
Fix ability to edit a discount before being synced

iamgabrielma commented:
Thanks for the review!

> The add discount screen is shown again, not the edit discount view.

Addressed [here](666b319628b4bcb71d2cb94061e624e30a4464a2) and [here](3bf84cd3dbfefe98a8e68f69835d2de78e8c0393). It was a combination of forgetting to change the text to "Edit Discount" when there's an existing one (first commit), and allowing a merchant to edit a discount before the remote sync has completed, the second commit redacts the view while is loading so now can only be edited when the sync is done.

iamgabrielma commented:
> * There's no gap between the edge of the screen and the segmented control

Fixed here e05f82ae95f4a9dce9d41f6ed140007356eca7cc

> * The product name is centred when it should be leading aligned

Fixed here a0650cc523afc76adfc33a21373bc8184d4ea4f7

> * The `×` symbol looks huge. It's from another PR but I think it'd be better if you pasted the multiplication symbol into the code and used a localized format string rather than using an image.

I'll address this one with other UI improvements here: https://github.com/woocommerce/woocommerce-ios/issues/10863
